pkg/manager/volumes/vol_compare_utils.go

```diff
@@ -0,0 +1,385 @@
+		return ""
+	}
+	return *v.StorageClassName
+}
+
+func (v *DesiredVolume) GetStorageSize() resource.Quantity {
+	return v.Size
+}
+
+type volCompareUtils struct {
+	deps *controller.Dependencies
+	sf   *selectorFactory
+}
+
+func newVolCompareUtils(deps *controller.Dependencies) *volCompareUtils {
+	return &volCompareUtils{
+		deps: deps,
+		sf:   MustNewSelectorFactory(),
+	}
+}
+
+type componentVolumeContext struct {
+	context.Context
+	tc     *v1alpha1.TidbCluster
+	status v1alpha1.ComponentStatus
+
+	shouldEvict bool
+
+	pods []*corev1.Pod
+	sts  *appsv1.StatefulSet
+
+	desiredVolumes []DesiredVolume
+}
+
+func (c *componentVolumeContext) ComponentID() string {
+	return fmt.Sprintf("%s/%s:%s", c.tc.GetNamespace(), c.tc.GetName(), c.status.MemberType())
+}
+
+func (u *volCompareUtils) BuildContextForTC(tc *v1alpha1.TidbCluster, status v1alpha1.ComponentStatus) (*componentVolumeContext, error) {
+	comp := status.MemberType()
+
+	ctx := &componentVolumeContext{
+		Context: context.TODO(),
+		tc:      tc,
+		status:  status,
+	}
+
+	vs, err := u.GetDesiredVolumes(tc, comp)
+	if err != nil {
+		return nil, err
+	}
+	ctx.desiredVolumes = vs
+
+	sts, err := u.getStsOfComponent(tc, comp)
+	if err != nil {
+		return nil, err
+	}
+
+	pods, err := u.getPodsOfComponent(tc, comp)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx.pods = pods
+	ctx.sts = sts
+	ctx.shouldEvict = comp == v1alpha1.TiKVMemberType
+
+	return ctx, nil
+}
+
+func (u *volCompareUtils) getPodsOfComponent(tc *v1alpha1.TidbCluster, mt v1alpha1.MemberType) ([]*corev1.Pod, error) {
+	selector, err := u.sf.NewSelector(tc.GetInstanceName(), mt)
+	if err != nil {
+		return nil, err
+	}
+
+	ns := tc.GetNamespace()
+
+	pods, err := u.deps.PodLister.Pods(ns).List(selector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list Pods: %w", err)
+	}
+
+	sort.Slice(pods, func(i, k int) bool {
+		a, b := pods[i].Name, pods[k].Name
+		if len(a) != len(b) {
+			return len(a) < len(b)
+		}
+		return a < b
+	})
+
+	return pods, nil
+}
+
+func (u *volCompareUtils) getStsOfComponent(cluster v1alpha1.Cluster, mt v1alpha1.MemberType) (*appsv1.StatefulSet, error) {
+	ns := cluster.GetNamespace()
+	stsName := controller.MemberName(cluster.GetName(), mt)
+
+	sts, err := u.deps.StatefulSetLister.StatefulSets(ns).Get(stsName)
+	if err != nil {
+		return nil, fmt.Errorf("get sts %s/%s failed: %w", ns, stsName, err)
+	}
+
+	return sts, nil
+}
+
+func (u *volCompareUtils) IsStatefulSetSynced(ctx *componentVolumeContext, sts *appsv1.StatefulSet, errOnNewVol bool) (bool, error) {
+	var newVolErr error = nil
+	for _, volTemplate := range sts.Spec.VolumeClaimTemplates {
+		volName := v1alpha1.StorageVolumeName(volTemplate.Name)
+		size := getStorageSize(volTemplate.Spec.Resources.Requests)
+		desired := getDesiredVolumeByName(ctx.desiredVolumes, volName)
+		if desired == nil {
+			// TODO: verify behaviour is correctly intended, this was previously being ignored. Now explicit error in
+			// pvc modifier..
```

Comment:
Removed the TODO, but I left this as a note to call out during review.
This function was originally used by existing pvc modifier here:
https://github.com/pingcap/tidb-operator/blob/899409c2f2858b6aa8e2f9a46db4607e32ee6fc4/pkg/manager/volumes/pvc_modifier.go#L201

We are now checking change in number of volumes and marking it as an error for pvc modifier (since it cannot handle it). Previously it would silently ignore it and allow other updates to continue. So this is a new behavior change I wanted to call out to make sure it's okay.